src/theory/symmetry_detect.cpp

```diff
@@ -0,0 +1,525 @@
+  map<Node, Node> subvar_to_expr;
+  // Partitions that have 2 or more variables
+  vector<Partition> new_partitions;
+  vector<Partition>::iterator part_it = partitions.begin();
+
+  // Collect singleton partitions: subvar_to_expr, subvar_to_var, and variables
+  while (part_it != partitions.end())
+  {
+    if (!(*part_it).d_var_to_subvar.empty()
+        && (*part_it).d_var_to_subvar.size() == 1)
+    {
+      vars.insert(((*part_it).d_var_to_subvar.begin())->first);
+      subvar_to_expr[((*part_it).d_var_to_subvar.begin())->second] =
+          (*part_it).d_term;
+      subvar_to_var[((*part_it).d_var_to_subvar.begin())->second] =
+          ((*part_it).d_var_to_subvar.begin())->first;
+    }
+    else if (!(*part_it).d_var_to_subvar.empty()
+             && (*part_it).d_var_to_subvar.size() >= 2)
+    {
+      map<Node, Node>::iterator var_to_subvar_it =
+          (*part_it).d_var_to_subvar.begin();
+
+      while (var_to_subvar_it != (*part_it).d_var_to_subvar.end())
+      {
+        vars.insert(var_to_subvar_it->first);
+        ++var_to_subvar_it;
+      }
+      // Only add partitions that have more than 1 variable
+      new_partitions.push_back(*part_it);
+    }
+    ++part_it;
+  }
+
+  // Save all partitions that have more than 1 variable
+  partitions = new_partitions;
+
+  // Do matches on singleton terms
+  if (!subvar_to_var.empty())
+  {
+    matches(partitions, subvar_to_var, subvar_to_expr);
+  }
+}
+
+void SymmetryDetect::collectChildren(Node node, vector<Node>& children)
+{
+  if (theory::quantifiers::TermUtil::isAssoc(node.getKind()))
+  {
+    for(const Node& n : node)
+    {
+      collectChildren(n, node.getKind(), children);
+    }
+  }
+  else
+  {
+    for(const Node& n : node)
+    {
+      children.push_back(n);
+    }
+  }
+}
+
+void SymmetryDetect::collectChildren(Node node, Kind k, vector<Node>& children)
+{
+  if (node.getKind() == k)
+  {
+    for(const Node& n : node)
+    {
+      collectChildren(n, k, children);
+    }
+  }
+  else
+  {
+    children.push_back(node);
+  }
+}
+
+void SymmetryDetect::PartitionTrie::getNewPartition(Partition& part)
+{
+  Trace("sym-dt") << "[sym-dt] Getting the new partitions from the leaves of "
+                     "the partition trie:"
+                  << endl;
+  if (!d_variables.empty())
+  {
+    vector<Node> vars;
+    vector<Node>::iterator var_it = d_variables.begin();
+    Node fresh_var =
+        NodeManager::currentNM()->mkSkolem("sym_bk", d_variables[0].getType());
+
+    while (var_it != d_variables.end())
+    {
+      vars.push_back(*var_it);
+      part.d_var_to_subvar[*var_it] = fresh_var;
+      ++var_it;
+    }
+    part.d_subvar_to_vars[fresh_var] = vars;
+  }
+  else
+  {
+    map<Node, PartitionTrie>::iterator part_it = d_children.begin();
+
+    while (part_it != d_children.end())
+    {
+      getNewPartition(part, part_it->second);
+      ++part_it;
+    }
+  }
+}
+
+void SymmetryDetect::PartitionTrie::getNewPartition(Partition& part,
```

Comment:
The code in the two `getNewPartition()` variants seems to be virtually identical. Any chance that we could deduplicate that code?

Reply:
In fact, the two functions are not identical at all. The first getNewPartition() is to check whether the current node of a partition trie is a leaf node `if (!d_variables.empty())`, but the second one is to check whether its children is a leaf node if `(!pt.d_variables.empty())`.

Reply:
Can we have the former just call:

getPartition(part, *this); 

?